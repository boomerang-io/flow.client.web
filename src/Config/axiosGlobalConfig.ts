// Setup global config for Axios requests so they work with ICP
import axios from "axios";
axios.defaults.withCredentials = true;
