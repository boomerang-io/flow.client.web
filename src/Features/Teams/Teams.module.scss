.content {
  overflow: auto;
  position: relative;
}

.createTeamTrigger {
  position: absolute;
  right: 2rem;
  top: 2rem;
}

.tableRow:hover {
  cursor: pointer;
}
