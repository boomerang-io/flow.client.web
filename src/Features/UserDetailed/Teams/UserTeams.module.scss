.container {
  padding: 1.5rem 2rem;

  :global .cds--bmrg-error-page {
    width: 100%;
  }

  :global .cds--structured-list {
    max-width: 100rem;
  }

  :global .cds--structured-list-td {
    line-height: 1rem;
    padding: 1rem 1rem;
  }
}

.actionsContainer {
  align-items: flex-start;
  display: flex;
  justify-content: space-between;
  margin-bottom: 2rem;
  width: 100%;

  :global .cds--search {
    max-width: 24rem;
  }
}

.leftActions {
  min-width: 24rem;
}

.featureDescription {
  color: var(--cds-text-01);
  font-size: 0.875rem;
  line-height: 1.125rem;
  margin-bottom: 1.375rem;
}

.teamCountText {
  color: var(--cds-text-01);
  font-size: 0.875rem;
  line-height: 1.125rem;
  margin-bottom: 0.5rem;
}

.viewTeamLink {
  color: var(--cds-interactive-01);
  cursor: pointer;
  font-size: 0.75rem;

  &[data-readonly="true"] {
    pointer-events: none;
    opacity: 0.7;
  }
}
