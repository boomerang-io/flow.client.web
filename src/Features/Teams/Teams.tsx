import React from "react";
import { Helmet } from "react-helmet";
import { useQuery } from "Hooks";
import { useFeature } from "flagged";
import { useHistory, useLocation, Route, Switch } from "react-router-dom";
import { Box } from "reflexbox";
import { Button, DataTable, DataTableSkeleton, Pagination, Search } from "@carbon/react";
import { CheckmarkFilled, Misuse } from "@carbon/react/icons";
import {
  ComposedModal,
  ErrorMessage,
  FeatureHeader as Header,
  FeatureHeaderTitle as HeaderTitle,
  FeatureHeaderSubtitle as HeaderSubtitle,
} from "@boomerang-io/carbon-addons-boomerang-react";
import TeamDetailed from "Features/TeamDetailed";
import EmptyState from "Components/EmptyState";
import moment from "moment";
import AddTeamContent from "./AddTeamContent";
import debounce from "lodash/debounce";
import queryString from "query-string";
import { isAccessibleKeyboardEvent } from "@boomerang-io/utils";
import { SortDirection } from "Constants";
import { AppPath, appLink, queryStringOptions, FeatureFlag } from "Config/appConfig";
import { serviceUrl } from "Config/servicesConfig";
import { ComposedModalChildProps, ModalTriggerProps, PaginatedTeamResponse } from "Types";
import styles from "./Teams.module.scss";

const TeamsContainer: React.FC = () => {
  return (
    <Switch>
      <Route path={AppPath.Team}>
        <TeamDetailed />
      </Route>
      <Route path={AppPath.TeamList}>
        <TeamList />
      </Route>
    </Switch>
  );
};

interface FeatureLayoutProps {
  handleSearchChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export default TeamsContainer;

const FeatureLayout: React.FC<FeatureLayoutProps> = ({ children, handleSearchChange }) => {
  return (
    <>
      <Helmet>
        <title>Teams</title>
      </Helmet>
      <Header
        includeBorder={false}
        header={
          <>
            <HeaderTitle style={{ margin: "0" }}>Teams</HeaderTitle>
            <HeaderSubtitle>View and manage teams</HeaderSubtitle>
          </>
        }
      />
      <Box p="2rem" className={styles.content}>
        <>
          <Box mb="1rem" maxWidth="20rem">
            <Search id="flow-teams" labelText="Search teams" placeholder="Search teams" onChange={handleSearchChange} />
          </Box>
          {children}
        </>
      </Box>
    </>
  );
};

const DEFAULT_ORDER = "DESC";
const DEFAULT_PAGE = 0;
const DEFAULT_LIMIT = 10;
const DEFAULT_SORT = "name";
const PAGE_SIZES = [DEFAULT_LIMIT, 20, 50, 100];

const TeamList: React.FC = () => {
  const history = useHistory();
  const location = useLocation();
  const cancelRequestRef = React.useRef<{} | null>();
  // TODO - make this read only
  const teamManagementEnabled = useFeature(FeatureFlag.TeamManagementEnabled);

  /**
   * Prepare queries and get some data
   */
  const {
    order = DEFAULT_ORDER,
    page = DEFAULT_PAGE,
    limit = DEFAULT_LIMIT,
    sort = DEFAULT_SORT,
  } = queryString.parse(location.search, queryStringOptions);

  const teamsUrlQuery = queryString.stringify({
    order,
    page,
    limit,
    sort,
  });

  const teamsUrl = serviceUrl.getTeams({ query: teamsUrlQuery });

  const {
    data: teamsData,
    error: teamsIsError,
    isLoading: teamsIsLoading,
  } = useQuery<PaginatedTeamResponse, string>(teamsUrl);

  const teamsQuery = useQuery(serviceUrl.getTeams({ query: location.search }));

  /**
   * Function that updates url search history to persist state
   * @param {object} query - all of the query params
   *
   */
  function updateHistorySearch({
    order = DEFAULT_ORDER,
    page = DEFAULT_PAGE,
    limit = DEFAULT_LIMIT,
    sort = DEFAULT_SORT,
    ...props
  }) {
    const queryStr = `?${queryString.stringify({ order, page, limit, sort, ...props })}`;
    history.push({ search: queryStr });
    return;
  }
  // eslint-disable-next-line
  const debouncedSearch = React.useCallback(
    debounce((query: string) => {
      updateHistorySearch({ query, page: 0 });
    }, 300),
    []
  );

  function handleSearchChange(e: React.ChangeEvent<HTMLInputElement>) {
    const query = e.target.value;
    debouncedSearch(query);
  }

  function handleNavigateToTeam(teamId: string) {
    history.push(appLink.team({ teamId }));
  }

  if (teamsQuery.isLoading) {
    return (
      <FeatureLayout handleSearchChange={handleSearchChange}>
        <DataTableSkeleton />
      </FeatureLayout>
    );
  }

  if (teamsQuery.isError) {
    return (
      <FeatureLayout handleSearchChange={handleSearchChange}>
        <ErrorMessage />
      </FeatureLayout>
    );
  }
  return (
    <FeatureLayout handleSearchChange={handleSearchChange}>
      {teamsData && teamManagementEnabled && (
        <ComposedModal
          composedModalProps={{ shouldCloseOnOverlayClick: true }}
          modalHeaderProps={{
            title: "Create Team",
            subtitle: `Scope your workflows and parameters to a team`,
          }}
          modalTrigger={({ openModal }: ModalTriggerProps) => (
            <Button
              iconDescription="Create new version"
              onClick={openModal}
              size="md"
              disabled={teamsIsError || teamsIsLoading}
              className={styles.createTeamTrigger}
            >
              Create Team
            </Button>
          )}
        >
          {({ closeModal }: ComposedModalChildProps) => {
            return (
              <AddTeamContent
                closeModal={closeModal}
                cancelRequestRef={cancelRequestRef}
                teamRecords={teamsData.content}
                currentQuery={location.search}
              />
            );
          }}
        </ComposedModal>
      )}
      <TeamListTable
        handleNavigateToTeam={handleNavigateToTeam}
        location={location}
        sort={sort}
        order={order}
        tableData={teamsQuery.data}
        updateHistorySearch={updateHistorySearch}
      />
    </FeatureLayout>
  );
};

const headers = [
  {
    header: "Name",
    key: "name",
    sortable: true,
  },
  {
    header: "Date Created",
    key: "creationDate",
    sortable: true,
  },
  {
    header: "# of Users",
    key: "users",
  },
  { header: "# of Workflows", key: "quotas" },
  { header: "Status", key: "status" },
];

interface TeamListTableProps {
  handleNavigateToTeam: Function;
  location: any;
  sort: string;
  order: string;
  tableData: {
    number: number;
    size: number;
    totalElements: number;
    content: any;
  };
  updateHistorySearch: Function;
}

function TeamListTable(props: TeamListTableProps) {
  const { number, size, totalElements, content } = props.tableData;
  const { TableContainer, Table, TableHead, TableRow, TableBody, TableCell, TableHeader } = DataTable;

  function handlePaginationChange({ page, pageSize }: { page: number; pageSize: number }) {
    props.updateHistorySearch({
      ...queryString.parse(props.location.search),
      page: page - 1, // We have to decrement by one to offset the table pagination adjustment
      limit: pageSize,
    });
  }

  function handleSort(e: any, { sortHeaderKey }: { sortHeaderKey: string }) {
    let order = "ASC";
    if (props.order === "ASC") {
      order = "DESC";
    }
    props.updateHistorySearch({ ...queryString.parse(props.location.search), sort: sortHeaderKey, order });
  }

  return content.length > 0 ? (
    <>
      <DataTable
        rows={content}
        headers={headers}
        render={({ rows, headers, getHeaderProps }: any) => (
          <TableContainer>
            <Table isSortable>
              <TableHead>
                <TableRow>
                  {headers.map((header: any) => (
                    <TableHeader
                      id={header.key}
                      {...getHeaderProps({
                        header,
                        isSortable: header.sortable,
                        onClick: handleSort,
                      })}
                      isSortHeader={props.sort === header.key}
                      sortDirection={props.order}
                    >
                      {header.header}
                    </TableHeader>
                  ))}
                </TableRow>
              </TableHead>
              <TableBody>
                {rows.map((row: any) => (
                  <TableRow
                    className={styles.tableRow}
                    key={row.id}
                    data-testid="user-list-table-row"
                    onClick={() => props.handleNavigateToTeam(row.id)}
                    onKeyDown={(e: React.SyntheticEvent) =>
                      isAccessibleKeyboardEvent(e) && props.handleNavigateToTeam(row.id)
                    }
                    tabIndex={-1}
                  >
                    {row.cells.map((cell: any, cellIndex: any) => {
                      if (cell.info.header === "status") {
                        return (
                          <TableCell key={cell.id} id={cell.id}>
                            {cell.value === "active" ? (
                              <CheckmarkFilled aria-label="Active" fill="green" />
                            ) : (
                              <Misuse aria-label="Inactive" fill="red" />
                            )}
                          </TableCell>
                        );
                      } else if (cell.info.header === "creationDate") {
                        return (
                          <TableCell key={cell.id}>
                            <time>{moment(cell.value).format("YYYY-MM-DD hh:mm A")}</time>
                          </TableCell>
                        );
                      } else if (cell.info.header === "quotas") {
                        return <TableCell key={cell.id}>{cell.value.currentWorkflowCount}</TableCell>;
                      }
                      return (
                        <TableCell key={cell.id}>
                          {Array.isArray(cell.value) ? cell.value.length : cell?.value ?? "---"}
                        </TableCell>
                      );
                    })}
                  </TableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        )}
      />
      <Pagination
        onChange={handlePaginationChange}
        page={number + 1}
        pageSize={size}
        pageSizes={PAGE_SIZES}
        totalItems={totalElements}
      />
    </>
  ) : (
    <EmptyState message={null} title="No teams found" />
  );
}
