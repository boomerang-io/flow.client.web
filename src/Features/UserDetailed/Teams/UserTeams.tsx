import React from "react";
import { Helmet } from "react-helmet";
import sortBy from "lodash/sortBy";
import {
  Search,
  StructuredListBody,
  StructuredListCell,
  StructuredListHead,
  StructuredListRow,
  StructuredListWrapper,
} from "@carbon/react";
import { Link } from "react-router-dom";
import EmptyState from "Components/EmptyState";
import { matchSorter as ms } from "match-sorter";
import { appLink } from "Config/appConfig";
import { FlowUser, FlowTeam } from "Types";
import styles from "./UserTeams.module.scss";

interface UserTeamsProps {
  user: FlowUser;
  teams: Array<FlowTeam>;
}

function UserTeams({ user, teams }: UserTeamsProps) {
  const [searchQuery, setSearchQuery] = React.useState("");
  const userTeams = teams ?? [];
  const filteredTeamsList = searchQuery ? ms(userTeams, searchQuery, { keys: ["name"] }) : userTeams;

  return (
    <section aria-label={`${user.name} Teams`} className={styles.container}>
      <Helmet>
        <title>{`Teams - ${user.name}`}</title>
      </Helmet>
      <section className={styles.actionsContainer}>
        <div className={styles.leftActions}>
          <p className={styles.featureDescription}>{`These are ${user.name}'s teams`}</p>
          <p className={styles.teamCountText}>
            Showing {filteredTeamsList.length} team{filteredTeamsList.length !== 1 ? "s" : ""}
          </p>
          <Search
            labelText="teams search"
            id="teams-search"
            placeholder="Search for a team"
            onChange={(e: React.FormEvent<HTMLInputElement>) => setSearchQuery(e.currentTarget.value)}
          />
        </div>
      </section>
      {filteredTeamsList.length > 0 ? (
        <StructuredListWrapper>
          <StructuredListHead>
            <StructuredListRow head>
              <StructuredListCell head>Name</StructuredListCell>
              <StructuredListCell head />
            </StructuredListRow>
          </StructuredListHead>
          <StructuredListBody>
            {sortBy(filteredTeamsList, "name").map((team) => (
              <StructuredListRow key={team.id}>
                <StructuredListCell>{team.name}</StructuredListCell>
                <StructuredListCell>
                  <Link
                    className={styles.viewTeamLink}
                    to={{
                      pathname: appLink.team({ teamId: team.id }),
                      state: { fromUser: { name: user.name, id: user.id } },
                    }}
                  >
                    View team
                  </Link>
                </StructuredListCell>
              </StructuredListRow>
            ))}
          </StructuredListBody>
        </StructuredListWrapper>
      ) : (
        <EmptyState />
      )}
    </section>
  );
}

export default UserTeams;
