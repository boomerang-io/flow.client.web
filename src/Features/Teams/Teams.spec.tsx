import React from "react";
import Teams from ".";
import { Route } from "react-router-dom";
import { waitFor, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { AppPath, appLink } from "Config/appConfig";
import { startApiServer } from "ApiServer";

let server:any;

beforeEach(() => {
  server = startApiServer();
});

afterEach(() => {
  server.shutdown();
});


describe("Teams --- Snapshot Test", () => {
  it("Capturing Snapshot of Teams", async () => {
    const { baseElement, findByText } = global.rtlContextRouterRender(
      <Route path={AppPath.TeamList}>
        <Teams />
      </Route>,
      { route: appLink.teamList() }
    );
    await findByText("WDC2 ISE QA");
    expect(baseElement).toMatchSnapshot();
    await waitFor(() => {});
  });
});

describe("Teams --- RTL", () => {
  test("Create new team", async () => {
    global.rtlContextRouterRender(
      <Route path={AppPath.TeamList}>
        <Teams />
      </Route>,
      { route: appLink.teamList() }
    );
    const createTeamButton = await screen.findByText(/^Create Team$/i);
    fireEvent.click(createTeamButton);
    expect(screen.getByText(/^Scope your workflows and parameters to a team$/i)).toBeInTheDocument();
    expect(screen.getByText(/^Save$/i)).toBeDisabled();
    const teamNameInput = screen.getByLabelText(/^Name$/i);
    userEvent.type(teamNameInput, "Test team");
    expect(screen.getByText(/^Save$/i)).toBeEnabled();
    fireEvent.click(screen.getByText(/^Save$/i));
    expect(await screen.findByText(/Test team/i)).toBeInTheDocument();
    await waitFor(() => {});
  });
});
