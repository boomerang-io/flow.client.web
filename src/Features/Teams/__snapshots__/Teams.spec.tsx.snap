// Vitest Snapshot v1

exports[`Teams --- Snapshot Test > Capturing Snapshot of Teams 1`] = `
<body
  id="app"
>
  <div>
    <header
      class="cds--bmrg-feature-header"
    >
      <section
        class="cds--bmrg-feature-header__content"
      >
        <div
          class="cds--bmrg-feature-header__header"
        >
          <h1
            class="cds--bmrg-feature-header-title"
            style="margin: 0px;"
          >
            Teams
          </h1>
          <p
            class="cds--bmrg-feature-header-title"
          >
            View and manage Flow teams
          </p>
        </div>
      </section>
    </header>
    <div
      class="_content_861edb css-1aaedcz"
    >
      <div
        class="css-2o893x"
      >
        <div
          aria-labelledby="flow-teams-search"
          class="cds--search cds--search--md"
          role="search"
        >
          <div
            class="cds--search-magnifier"
          >
            <svg
              aria-hidden="true"
              class="cds--search-magnifier-icon"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M15,14.3L10.7,10c1.9-2.3,1.6-5.8-0.7-7.7S4.2,0.7,2.3,3S0.7,8.8,3,10.7c2,1.7,5,1.7,7,0l4.3,4.3L15,14.3z M2,6.5 C2,4,4,2,6.5,2S11,4,11,6.5S9,11,6.5,11S2,9,2,6.5z"
              />
            </svg>
          </div>
          <label
            class="cds--label"
            for="flow-teams"
            id="flow-teams-search"
          >
            Search teams
          </label>
          <input
            autocomplete="off"
            class="cds--search-input"
            id="flow-teams"
            placeholder="Search teams"
            role="searchbox"
            type="text"
            value=""
          />
          <button
            aria-label="Clear search input"
            class="cds--search-close cds--search-close--hidden"
            title="Clear search input"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M24 9.4L22.6 8 16 14.6 9.4 8 8 9.4 14.6 16 8 22.6 9.4 24 16 17.4 22.6 24 24 22.6 17.4 16 24 9.4z"
              />
            </svg>
          </button>
        </div>
      </div>
      <button
        class="_createTeamTrigger_861edb cds--btn cds--btn--md cds--btn--primary"
        tabindex="0"
        type="button"
      >
        Create Team
      </button>
      <div
        class="cds--data-table-container"
      >
        <div
          class="cds--data-table-content"
        >
          <table
            class="cds--data-table cds--data-table--sort"
          >
            <thead>
              <tr>
                <th
                  aria-sort="none"
                  id="name"
                  scope="col"
                >
                  <div
                    id="table-sort-4"
                    style="display: none;"
                  >
                    Click to sort rows by Name header in ascending order
                  </div>
                  <button
                    aria-describedby="table-sort-4"
                    class="cds--table-sort"
                    type="button"
                  >
                    <span
                      class="cds--table-sort__flex"
                    >
                      <div
                        class="cds--table-header-label"
                      >
                        Name
                      </div>
                      <svg
                        aria-hidden="true"
                        class="cds--table-sort__icon"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M16 4L6 14 7.41 15.41 15 7.83 15 28 17 28 17 7.83 24.59 15.41 26 14 16 4z"
                        />
                      </svg>
                      <svg
                        aria-hidden="true"
                        class="cds--table-sort__icon-unsorted"
                        fill="currentColor"
                        focusable="false"
                        height="20"
                        preserveAspectRatio="xMidYMid meet"
                        viewBox="0 0 32 32"
                        width="20"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M27.6 20.6L24 24.2 24 4 22 4 22 24.2 18.4 20.6 17 22 23 28 29 22zM9 4L3 10 4.4 11.4 8 7.8 8 28 10 28 10 7.8 13.6 11.4 15 10z"
                        />
                      </svg>
                    </span>
                  </button>
                </th>
                <th
                  id="users"
                  scope="col"
                >
                  <div
                    class="cds--table-header-label"
                  >
                    # of Users
                  </div>
                </th>
                <th
                  id="workflows"
                  scope="col"
                >
                  <div
                    class="cds--table-header-label"
                  >
                    # of Workflows
                  </div>
                </th>
                <th
                  id="isActive"
                  scope="col"
                >
                  <div
                    class="cds--table-header-label"
                  >
                    Active
                  </div>
                </th>
              </tr>
            </thead>
            <tbody
              aria-live="polite"
            >
              <tr
                class="_tableRow_861edb"
                data-testid="user-list-table-row"
                tabindex="-1"
              >
                <td>
                  WDC2 ISE Dev
                </td>
                <td>
                  2
                </td>
                <td>
                  ---
                </td>
                <td
                  id="5e7cccb94bbc6e0001c51773:isActive"
                >
                  <div
                    class="cds--form-item cds--checkbox-wrapper"
                  >
                    <input
                      checked=""
                      class="cds--checkbox"
                      id="check-5e7cccb94bbc6e0001c51773:isActive"
                      type="checkbox"
                    />
                    <label
                      class="cds--checkbox-label"
                      for="check-5e7cccb94bbc6e0001c51773:isActive"
                    >
                      <span
                        class="cds--checkbox-label-text cds--visually-hidden"
                        dir="auto"
                      >
                        checkbox
                      </span>
                    </label>
                  </div>
                </td>
              </tr>
              <tr
                class="_tableRow_861edb"
                data-testid="user-list-table-row"
                tabindex="-1"
              >
                <td>
                  WDC2 ISE QA
                </td>
                <td>
                  1
                </td>
                <td>
                  ---
                </td>
                <td
                  id="6e7cccb94bbc6e0001c51773:isActive"
                >
                  <div
                    class="cds--form-item cds--checkbox-wrapper"
                  >
                    <input
                      checked=""
                      class="cds--checkbox"
                      id="check-6e7cccb94bbc6e0001c51773:isActive"
                      type="checkbox"
                    />
                    <label
                      class="cds--checkbox-label"
                      for="check-6e7cccb94bbc6e0001c51773:isActive"
                    >
                      <span
                        class="cds--checkbox-label-text cds--visually-hidden"
                        dir="auto"
                      >
                        checkbox
                      </span>
                    </label>
                  </div>
                </td>
              </tr>
              <tr
                class="_tableRow_861edb"
                data-testid="user-list-table-row"
                tabindex="-1"
              >
                <td>
                  IBM Services Engineering
                </td>
                <td>
                  ---
                </td>
                <td>
                  3
                </td>
                <td
                  id="5e3a35ad8c222700018ccd39:isActive"
                >
                  <div
                    class="cds--form-item cds--checkbox-wrapper"
                  >
                    <input
                      class="cds--checkbox"
                      id="check-5e3a35ad8c222700018ccd39:isActive"
                      type="checkbox"
                    />
                    <label
                      class="cds--checkbox-label"
                      for="check-5e3a35ad8c222700018ccd39:isActive"
                    >
                      <span
                        class="cds--checkbox-label-text cds--visually-hidden"
                        dir="auto"
                      >
                        checkbox
                      </span>
                    </label>
                  </div>
                </td>
              </tr>
              <tr
                class="_tableRow_861edb"
                data-testid="user-list-table-row"
                tabindex="-1"
              >
                <td>
                  IBM Services Essentials
                </td>
                <td>
                  ---
                </td>
                <td>
                  1
                </td>
                <td
                  id="5e3a35ad8c222700018ccd38:isActive"
                >
                  <div
                    class="cds--form-item cds--checkbox-wrapper"
                  >
                    <input
                      class="cds--checkbox"
                      id="check-5e3a35ad8c222700018ccd38:isActive"
                      type="checkbox"
                    />
                    <label
                      class="cds--checkbox-label"
                      for="check-5e3a35ad8c222700018ccd38:isActive"
                    >
                      <span
                        class="cds--checkbox-label-text cds--visually-hidden"
                        dir="auto"
                      >
                        checkbox
                      </span>
                    </label>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
      <div
        class="cds--pagination cds--pagination--md"
      >
        <div
          class="cds--pagination__left"
        >
          <label
            class="cds--pagination__text"
            for="cds-pagination-select-id-8"
            id="cds-pagination-select-id-8-count-label"
          >
            Items per page:
          </label>
          <div
            class="cds--form-item cds--select__item-count"
          >
            <div
              class="cds--select cds--select--inline"
            >
              <div
                class="cds--select-input--inline__wrapper"
              >
                <div
                  class="cds--select-input__wrapper"
                >
                  <select
                    class="cds--select-input"
                    id="cds-pagination-select-id-8"
                  >
                    <option
                      class="cds--select-option"
                      value="10"
                    >
                      10
                    </option>
                    <option
                      class="cds--select-option"
                      value="20"
                    >
                      20
                    </option>
                    <option
                      class="cds--select-option"
                      value="50"
                    >
                      50
                    </option>
                    <option
                      class="cds--select-option"
                      value="100"
                    >
                      100
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="cds--select__arrow"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <span
            class="cds--pagination__text cds--pagination__items-count"
          >
            1–2 of 2 items
          </span>
        </div>
        <div
          class="cds--pagination__right"
        >
          <div
            class="cds--form-item cds--select__page-number"
          >
            <div
              class="cds--select cds--select--inline"
            >
              <label
                class="cds--label cds--visually-hidden"
                for="cds-pagination-select-id-8-right"
              >
                Page number, of 1 pages
              </label>
              <div
                class="cds--select-input--inline__wrapper"
              >
                <div
                  class="cds--select-input__wrapper"
                >
                  <select
                    class="cds--select-input"
                    id="cds-pagination-select-id-8-right"
                  >
                    <option
                      class="cds--select-option"
                      value="1"
                    >
                      1
                    </option>
                  </select>
                  <svg
                    aria-hidden="true"
                    class="cds--select__arrow"
                    fill="currentColor"
                    focusable="false"
                    height="16"
                    preserveAspectRatio="xMidYMid meet"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M8 11L3 6 3.7 5.3 8 9.6 12.3 5.3 13 6z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <span
            class="cds--pagination__text"
          >
            of 1 page
          </span>
          <div
            class="cds--pagination__control-buttons"
          >
            <span
              class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top cds--tooltip cds--icon-tooltip"
            >
              <button
                aria-labelledby="tooltip-9"
                class="cds--btn--icon-only cds--pagination__button cds--pagination__button--backward cds--pagination__button--no-index cds--btn cds--btn--md cds--btn--ghost cds--btn--disabled"
                disabled=""
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M20 24L10 16 20 8z"
                  />
                </svg>
              </button>
              <span
                aria-hidden="true"
                class="cds--popover"
                id="tooltip-9"
                role="tooltip"
              >
                <span
                  class="cds--popover-content cds--tooltip-content"
                >
                  Previous page
                </span>
                <span
                  class="cds--popover-caret"
                />
              </span>
            </span>
            <span
              class="cds--popover-container cds--popover--caret cds--popover--high-contrast cds--popover--top-right cds--tooltip cds--icon-tooltip"
            >
              <button
                aria-labelledby="tooltip-11"
                class="cds--btn--icon-only cds--pagination__button cds--pagination__button--forward cds--pagination__button--no-index cds--btn cds--btn--md cds--btn--ghost cds--btn--disabled"
                disabled=""
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  focusable="false"
                  height="16"
                  preserveAspectRatio="xMidYMid meet"
                  viewBox="0 0 32 32"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M12 8L22 16 12 24z"
                  />
                </svg>
              </button>
              <span
                aria-hidden="true"
                class="cds--popover"
                id="tooltip-11"
                role="tooltip"
              >
                <span
                  class="cds--popover-content cds--tooltip-content"
                >
                  Next page
                </span>
                <span
                  class="cds--popover-caret"
                />
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="cds--bmrg-modal-portal"
  />
</body>
`;
